src/pluginutils.c

```diff
@@ -514,4 +513,57 @@ void plugin_builder_connect_signals(GeanyPlugin *plugin,
 }
 
 
+/** Add additional data that corresponds to the plugin.
+ *
+ * @p pdata is the pointer going to be passed to the individual plugin callbacks
+ * of GeanyPlugin::funcs. When the  plugin is cleaned up, @p free_func is invoked for the data,
+ * which connects the data to the time the plugin is enabled.
+ *
+ * One intended use case is to set GObjects as data and have them destroyed automatically
+ * by passing g_object_unref() as @a free_func, so that member functions can be used
+ * for the @ref GeanyPluginFuncs (via wrappers) but you can set completely custom data.
+ *
+ * Be aware that this can only be called once and only by plugins registered via
+ * @ref geany_plugin_register(). So-called legacy plugins cannot not use this function.
+ *
+ * @note This function must not be called if the plugin was registered with
+ * geany_plugin_register_full().
```

Comment:
Maybe mention that this should basically *never* be called from inside `geany_load_module()`?  Because the destroy *will* be called at plugin unload if the plugin never was enabled.  This does make sense, as we have to call the free func at some point, but it might be unexpected (e.g. @codebrainz's usage would break).  Actually it might even make sense to disallow calling `geany_plugin_set_data()` directly form `load_module()` altogether.

Alternatively, explain that if called from `geany_load_module()` and the plugin is not enabled, it will still be called once (but that might just be more confusing, and doesn't provide anything more than `register_full()`).